pool/object.go

```diff
@@ -42,17 +81,57 @@ func (p *objectPool) Init(alloc Allocator) {
 func (p *objectPool) Get() interface{} {
 	select {
 	case v := <-p.values:
+		p.trySetGauges()
+		if p.refillLowWatermark > 0 && len(p.values) < p.refillLowWatermark {
+			p.tryFill()
+		}
 		return v
 	default:
+		p.trySetGauges()
+		if p.refillLowWatermark > 0 {
+			p.tryFill()
+		}
+		p.metrics.getOnEmpty.Inc(1)
 		return p.alloc()
 	}
 }
 
 func (p *objectPool) Put(obj interface{}) {
 	select {
 	case p.values <- obj:
+		p.trySetGauges()
 		return
 	default:
+		p.trySetGauges()
+		p.metrics.putOnFull.Inc(1)
+		return
+	}
+}
+
+func (p *objectPool) trySetGauges() {
+	if time.Now().UnixNano()%sampleObjectPoolLengthEvery == 0 {
+		p.setGauges()
+	}
+}
+
+func (p *objectPool) setGauges() {
+	p.metrics.free.Update(int64(len(p.values)))
+	p.metrics.total.Update(int64(p.size))
+}
+
+func (p *objectPool) tryFill() {
+	if !atomic.CompareAndSwapInt64(&p.filling, 0, 1) {
 		return
 	}
+	go func() {
+		full := false
+		for !full {
+			select {
+			case p.values <- p.alloc():
```

Comment:
I wonder if we always want to replenish the entire pool as opposed to say 25% of the pool size, because if the demand for more objects is transient, we might allocate more than necessary